### ISSUE TITLE
Function taking variable type in

### ISSUE THREAD

**Issue by vron** [ronniec95/xladd-derive#5]
Hi Again,

Sorry to keep bothering you - am I correct in uderstanding that the derive function (which provide nice Result<> wrapping!) does not support any input type that holds a variable typed input array / matrix?

I.e. if i in Excel do =my_test(A1:B8)

In the underlying crate I could then define a function that would give me a single LPXLOPER12 / Variant in where i Could iterate over each matrix element and do different things depending on if they contain strings or numbers e.g. (but then without the error wrapping)

I naively tried to derive such a function using:

    #[xl_func(category = "test")]
    fn ra_json_object(val: Variant) -> Result<Variant, Box<dyn std::error::Error>> {
    
Without any luck

**Comment by ronniec95** [comment]
Just thinking about it, if you want to take a variant type then you can always drop down to xladd's raw interface.

That's where I originally started but with null values and conversions, my code got really messy (lots of it/else statements) which is why I wrote the xladd-derive macro.

On my todo list is Optional arguments represented by Option<f64> so if there is a need for that, let me know

**Comment by vron** [comment]
Thanks for clarification,

Yes - i can drop down - but I'd have to write the wrapper to handle the results (not much work admitedly) - I just though that a Variant as input would also be reasonable (since it's safe . the non strongly typed part is protected) - but thats fine - your crate your call :-)

As for options, always nice but would not help me now - i need json input to send so was planning to effectively write a add-in function that takes two columns of "field names" in col1 and "values" in col2 that i then merge to a json string to return / send  (the values in the json struct needs to be serialized diffrently depending on if they are numbers or strings or bools etc.)

However - your string workarkound should work very well - a bit unecissary to go float - string - float but not a problem for this case.

thanks

**Comment by vron** [comment]
Fair point on ndarray! Its an internal project im affraid so im nit allowed to share details :-(
(But no not doing matrix operations it was just a easy example)

**Comment by ronniec95** [comment]
Discussion complete
